### PR TITLE
Olwa dev 3

### PR DESCRIPTION
**Pull request by oliver-walker-hms**

### CODE REVIEW

File: src/com/hms/flexyosisoftconnector/OSIsoftServer.java [hunk 1]
```diff
@@ -212,231 +281,64 @@ public int setTagWebId(Tag tag) throws JSONException {
...
           if (res == NO_ERROR) {
             // Parse the JSON response and retrieve the JSON Array of items
-            JsonT = new JSONTokener(FileReader.readFile("file://" + responseFilename));
-            requestResponse = new JSONObject(JsonT);
-            if (requestResponse.has("Items")) {
-              items = requestResponse.getJSONArray("Items");
+            try {
+              JsonT = new JSONTokener(FileAccessManager.readFileToString(responseFilename));
+            } catch (IOException e) {
+              e.printStackTrace();
```
Comment (alexjhawk) [inline src/com/hms/flexyosisoftconnector/OSIsoftServer.java:287]:
```suggestion
              Logger.LOG_EXCEPTION(e);
```

File: src/com/hms/flexyosisoftconnector/RestFileServer.java [hunk 2]
```diff
@@ -16,7 +17,7 @@
  * <p>Class object for a RestFileServer. Listens on a socket for a file, if received the file is
  * saved.
  *
- * <p>HMS Networks Inc. Solution Center
```
Comment (alexjhawk) [inline src/com/hms/flexyosisoftconnector/RestFileServer.java]: unrelated change to commit

File: src/com/hms/flexyosisoftconnector/Main.java [hunk 3]
```diff
@@ -20,32 +25,22 @@
...
 
-  /** Configuration of PI connector */
-  static OSIsoftConfig piConfig;
-
   /** PI Server management object */
   static OSIsoftServer piServer;
 
-  /** Current available JVM memory */
-  static long AvailableMemory;
+  static PayloadManager payloadMngr;
```
Comment (alexjhawk) [inline src/com/hms/flexyosisoftconnector/Main.java:37]: javadoc

File: src/com/hms/flexyosisoftconnector/Main.java [hunk 4]
```diff
@@ -54,72 +49,75 @@ public static void main(String[] args) {
...
+        Thread.sleep(threadWaitMS);
+      } catch (InterruptedException e) {
+        e.printStackTrace();
+      }
+    }
+  }
 
-        // Update the last update time
-        lastUpdateTimeMs = currentTimeMs;
+  private static void getNewPayloadToSend() {
```
Comment (alexjhawk) [inline src/com/hms/flexyosisoftconnector/Main.java:100]: javadoc getNewPayloadToSend()

File: src/com/hms/flexyosisoftconnector/DataPoster.java [hunk 5]
```diff
@@ -1,168 +1,64 @@
...
         }
       }
 
-      // Tell the JVM that it should garbage collect soon
-      System.gc();
+      // thread finished
+      Thread.yield();
+      try {
+        int oneSecond = 1000;
+        Thread.sleep(oneSecond);
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/DataPoster.java:57]:
```suggestion
        final int sleepTimeMs = 1000;
        Thread.sleep(sleepTimeMs);
```
putting the value in the variable name is not helpful and leads to more code maintenance if you need to change it.

File: src/com/hms/flexyosisoftconnector/TagPayload.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
+package com.hms.flexyosisoftconnector;
+
+import com.hms_networks.americas.sc.datapoint.DataPoint;
+
+/**
+ * This class holds the incomplete payload as well as one of the data point objects.
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/TagPayload.java:6]:
```suggestion
 * This class holds the incomplete payload as well as one of the datapoint objects.
```

File: src/com/hms/flexyosisoftconnector/OSIsoftServer.java [hunk 7]
```diff
@@ -53,47 +52,46 @@
...
-    flexyName = name;
-    targetURL = "https://" + serverIP + "/piwebapi/";
+  /**
+   * Constructs the OSIsoftServer object.
+   *
+   * @param deviceName the name of the device
+   */
+  public OSIsoftServer(String deviceName) {
+    flexyName = deviceName;
+    targetURL = "https://" + OSIsoftConfig.getServerIP() + "/piwebapi/";
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/OSIsoftServer.java:67]: Why are you saving this everywhere and not just accessing it from the config?

File: src/com/hms/flexyosisoftconnector/OSIsoftServer.java [hunk 8]
```diff
@@ -53,47 +52,46 @@
...
+  /**
+   * Makes an HTTP request against the OSIsoft server and validates the response code.
+   *
+   * @param CnxParam Request parameters
+   * @param Method Request method
+   * @param Headers Request headers
+   * @param TextFields Request Payload
+   * @param FileFields File location for response
+   * @param FileName File name for response
+   * @return returns the HTTP response code
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/OSIsoftServer.java:92]:
```suggestion
   * @return The HTTP response code

```

File: src/com/hms/flexyosisoftconnector/OSIsoftServer.java [hunk 9]
```diff
@@ -105,6 +103,7 @@ public static int RequestHTTPS(
     int res = NO_ERROR;
 
     try {
+      Logger.LOG_INFO("request is: " + CnxParam);
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/OSIsoftServer.java:106]:
```suggestion
      Logger.LOG_INFO("Request is: " + CnxParam);

```

File: src/com/hms/flexyosisoftconnector/Main.java [hunk 10]
```diff
@@ -1,136 +1,123 @@
...
-  /**
-   * Minimum amount of free memory before data trimming occurs
-   */
-  static final long MINIMUM_MEMORY = 5000000;
 
-  /**
-   * Number of tags to trim when memory is low
-   */
-  static final int TAGS_TO_TRIM = 2;
+  static final int MINOR_VERSION = 0;
```
[the remark this answers is not in the document]
Comment (alexjhawk) [reply src/com/hms/flexyosisoftconnector/Main.java:26]: not fixed

**Comment by oliver-walker-hms** [comment]
I'll split these commits up a lot more after addressing all the current comments

**Comment by oliver-walker-hms** [comment]
@TomKimsey and @alexjhawk  I've gone ahead and reworked these commits to split them up. Do either of you see any other large areas of code that need to be in a separate commit? My only thoughts are maybe all the javadocs but that seems like overkill and not worth the effort to me. OSIsoft Server is still messy

File: src/com/hms/flexyosisoftconnector/OSIsoftServer.java (hunk 7 above)
Comment (oliver-walker-hms) [reply src/com/hms/flexyosisoftconnector/OSIsoftServer.java]: yeah I had missed changing this when you mentioned it on the other PR. I moved a couple of things from the OSIsoft server to the config class when I changed this

**Comment by oliver-walker-hms** [comment]
some commits need new names, this is not ready for review yet

File: src/com/hms/flexyosisoftconnector/Main.java [hunk 11]
```diff
@@ -21,129 +26,106 @@
...
+    do {
+      Logger.LOG_INFO("Initializing tags");
+      res = piServer.initTags();
 
-      AvailableMemory = Runtime.getRuntime().freeMemory();
-      currentTimeMs = System.currentTimeMillis();
+    } while (res != OSIsoftServer.NO_ERROR);
 
-      if ((currentTimeMs - lastUpdateTimeMs) >= piConfig.getCycleTimeMs()) {
+    Logger.LOG_INFO("got init");
```
Comment (alexjhawk) [inline src/com/hms/flexyosisoftconnector/Main.java:80]: "got init" doesn't seem descriptive.

Comment (oliver-walker-hms) [reply src/com/hms/flexyosisoftconnector/Main.java:80]: oops, this was for testing something

File: src/com/hms/flexyosisoftconnector/Main.java [hunk 12]
```diff
@@ -21,129 +26,106 @@
...
+
+        // build a payload out of data points
+        payloadMngr.appendDataPointToPayLoad(data);
+      }
     }
-    return res;
   }
 
   /**
-   * Sets the directory that the Ewon  uses to check for SSL Certificates
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Main.java]: do comment changes belong in this commit?

File: src/com/hms/flexyosisoftconnector/Main.java [hunk 13]
```diff
@@ -163,7 +145,7 @@ private static void setCertificatePath(String path) {
   }
 
   /**
-   * Sets the http timeouts Note: This changes the Ewon 's global HTTP timeouts and stores these
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Main.java]: comment change

File: src/com/hms/flexyosisoftconnector/Main.java [hunk 14]
```diff
@@ -183,9 +165,7 @@ private static void setHttpTimeouts() {
         needsSave = true;
       }
 
-      if (needsSave) {
-        SCB.saveBlock(true);
-      }
+      if (needsSave) SCB.saveBlock(true);
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Main.java]: why was this changed to inline?

File: src/com/hms/flexyosisoftconnector/Main.java [hunk 15]
```diff
@@ -15,9 +15,10 @@
  */
 public class Main {
 
-  /** Application Version Numbers */
-  static final int MAJOR_VERSION = 1;
+  /** Application Major Version Number */
+  static final int MAJOR_VERSION = 2;
 
+  /** Application Minor Version Number */
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Main.java]: one commit for changing the number, another for changing the comments.

For this commit the message is "edited Version numbers" so i would expect the values to change. But the commit message does not address the comments changing.

File: src/com/hms/flexyosisoftconnector/Main.java [hunk 16]
```diff
@@ -1,14 +1,19 @@
...
+import com.hms_networks.americas.sc.historicaldata.HistoricalDataQueueManager;
+import com.hms_networks.americas.sc.json.JSONException;
+import com.hms_networks.americas.sc.logging.Logger;
+import com.hms_networks.americas.sc.taginfo.TagInfoManager;
 
 /**
  * Ewon Flexy java demo for OSIsoft Server
  *
- * <p>This demo reads multiple tag values from an Ewon  Flexy IO Server and POSTs them to an OSIsoft
+ * <p>This demo reads multiple tag values from an Ewon Flexy IO Server and POSTs them to an OSIsoft
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Main.java]: comments changing ?

File: src/com/hms/flexyosisoftconnector/Main.java [hunk 17]
```diff
@@ -21,129 +26,106 @@
...
-
   /** Default name of a new Ewon */
   static final String DEFAULT_EWON_NAME = "eWON";
 
   /** Filename of connector config file */
   static final String CONNECTOR_CONFIG_FILENAME = "/usr/ConnectorConfig.json";
 
-  /** Configuration of PI connector */
-  static OSIsoftConfig piConfig;
-
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/Main.java]: Are these changes really related to using the new queue? Seems like there is more going on in this commit

File: src/com/hms/flexyosisoftconnector/OSIsoftConfig.java [hunk 18]
```diff
@@ -10,44 +15,76 @@
...
+   * Encode: "username:password"
+   */
   private static String piServerLogin;
 
-  // Path to the directory containing your OSIsoft server's certificate
+  /** Path to the directory containing your OSIsoft server's certificate */
   private static String ewonCertificatePath;
 
-  // Update rate for all tags in milliseconds
+  /** Update rate for all tags in milliseconds */
```
[the remark this answers is not in the document]
Comment (oliver-walker-hms) [reply src/com/hms/flexyosisoftconnector/OSIsoftConfig.java]: fixed

File: src/com/hms/flexyosisoftconnector/TagPayload.java [hunk 19]
```diff
@@ -0,0 +1,37 @@
+package com.hms.flexyosisoftconnector;
+
+import com.hms_networks.americas.sc.datapoint.DataPoint;
+
+/**
+ * This class holds the incomplete payload as well as one of the datapoint objects.
+ *
+ * @author HMS Networks Inc. Solution Center
+ */
+public class TagPayload {
```
[the remark this answers is not in the document]
Comment (oliver-walker-hms) [reply src/com/hms/flexyosisoftconnector/TagPayload.java:10]: string builder is used to construct the string, and this is only set once when the payload string is complete. The size limits are set in the OSIsoftJsonPayload class to be 200 data points worth of payload. I will swap this to string builder lite since it may be huge

[the remark this answers is not in the document]
Comment (TomKimsey) [reply src/com/hms/flexyosisoftconnector/TagPayload.java:10]: Then why is there an add method and not just a method like

  public void setPayload(String newPayload) {
    payload = newPayload;
  }

[the remark this answers is not in the document]
Comment (TomKimsey) [reply src/com/hms/flexyosisoftconnector/TagPayload.java:10]: Is this class just for one datapoint? if so I think the class description should be revised.

File: src/com/hms/flexyosisoftconnector/DataPoster.java [hunk 20]
```diff
@@ -1,168 +1,64 @@
...
 
-      // Tell the JVM that it should garbage collect soon
-      System.gc();
+      // thread finished
+      Thread.yield();
+      try {
+        final int sleepTimeMs = 1000;
+        Thread.sleep(sleepTimeMs);
+      } catch (InterruptedException e) {
+        Logger.LOG_SERIOUS("[DataPoster] Unable to sleep connected DataPoster thread!");
```
[the remark this answers is not in the document]
Comment (oliver-walker-hms) [reply src/com/hms/flexyosisoftconnector/DataPoster.java:59]: switched to remove since all the others do not have it

File: src/com/hms/flexyosisoftconnector/TagPayload.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (oliver-walker-hms) [reply src/com/hms/flexyosisoftconnector/TagPayload.java:10]: oh my mistake, yeah this did have a few chunks added in. forgot that i switched it.  osisoft json payload builds a new part of the payload and gives it to this class. it was just to keep the tag information coupled with the payload, so when the tag is first encountered it sets the data point once. Some information stored in the data point object is needed later on. It is for a single tag and a single payload, many data points. But yes, it should be string builder instead of string

[the remark this answers is not in the document]
Comment (oliver-walker-hms) [reply src/com/hms/flexyosisoftconnector/TagPayload.java:10]: I will make that change after everything else is done

File: src/com/hms/flexyosisoftconnector/OSIsoftServer.java [hunk 21]
```diff
@@ -162,31 +176,30 @@ public int setTagWebId(Tag tag) throws JSONException {
...
-      }
+      if (requestResponse.has("Items")) items = requestResponse.getJSONArray("Items");
 
       if (items.length() > 0) {
         // tag exists
-        tag.setWebID(items.getJSONObject(0).getString("WebId"));
+        tagWebID = items.getJSONObject(0).getString("WebId");
       } else {
         // tag does not exist and must be created
+        String payload = PayloadBuilder.buildNewPointBody(tag); // need datapoint? ////
```
Comment (alexjhawk) [inline src/com/hms/flexyosisoftconnector/OSIsoftServer.java:186]: leftover from debug?
```suggestion
        String payload = PayloadBuilder.buildNewPointBody(tag); 
```

File: src/com/hms/flexyosisoftconnector/OSIsoftServer.java [hunk 22]
```diff
@@ -162,31 +176,30 @@ public int setTagWebId(Tag tag) throws JSONException {
...
-                postHeaders,
-                buildNewPointBody(tagName, tag.getDataType()),
-                "",
-                responseFilename);
+                url, "Post", OSIsoftConfig.getPostHeaders(), payload, "", responseFilename);
 
         if (res == NO_ERROR) {
           // The WebID is sent back in the headers of the previous post
           // however, there is no mechanism currently to retrieve it so
           // another request must be issued.
```
Comment (alexjhawk) [inline src/com/hms/flexyosisoftconnector/OSIsoftServer.java]:
```suggestion
          /* The WebID is sent back in the headers of the previous post
              however, there is no mechanism currently to retrieve it so
              another request must be issued*/.
```
multiline comments should use block comment syntax instead of single comments

File: src/com/hms/flexyosisoftconnector/Main.java [hunk 23]
```diff
@@ -53,24 +61,65 @@ public static void main(String[] args) {
...
+    DataPoster dataThread = new DataPoster(payloadMngr);
+    dataThread.start();
+
+    while (true) {
+      getNewPayloadToSend();
+
+      // pause to help the flexy with the infinite loop
+      Thread.yield();
+      try {
+        int threadWaitMS = 500;
```
Comment (alexjhawk) [inline src/com/hms/flexyosisoftconnector/Main.java:89]:
```suggestion
        final int threadWaitMS = 500;
```

File: src/com/hms/flexyosisoftconnector/Main.java [hunk 24]
```diff
@@ -53,24 +61,65 @@ public static void main(String[] args) {
...
+        Thread.sleep(threadWaitMS);
+      } catch (InterruptedException e) {
+        e.printStackTrace();
```
Comment (alexjhawk) [inline src/com/hms/flexyosisoftconnector/Main.java:92]:
```suggestion
        Logger.LOG_EXCEPTION(e);
```

File: src/com/hms/flexyosisoftconnector/PayloadBuilder.java [hunk 25]
```diff
@@ -340,7 +340,7 @@ private static String getFlexyName() {
    * @return Returns a string representation of the constructed json body.
    */
   public static String buildNewPointBody(TagInfo tag) {
-    String type;
+    String type = "null";
```
[the remark this answers is not in the document]
Comment (TomKimsey) [reply src/com/hms/flexyosisoftconnector/PayloadBuilder.java:343]: This comment was not responded to

File: src/com/hms/flexyosisoftconnector/OsisoftJsonPayload.java [hunk 26]
```diff
@@ -88,6 +88,8 @@
   /** Error message to display on unknown communication type. */
   String comsErrMsg = "Unknown communication type given. Unable to build payload.";
 
+  public static int MAX_PAYLOAD_NUM_CHARACTERS = 0;
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/OsisoftJsonPayload.java:91]: The max is zero? Should this just be 
```suggestion
  public static int MAX_PAYLOAD_NUM_CHARACTERS;
```

Comment (oliver-walker-hms) [reply src/com/hms/flexyosisoftconnector/OsisoftJsonPayload.java:91]: I set that to 0 because TagPayload uses it in it's constructor which is called after the  OSIsoft JSON payload is constructed and that gets set to a valid value but there were still warnings about it having the potential to be uninitialized in TagPayload constructor. Some better error handling could probably be added there

File: src/com/hms/flexyosisoftconnector/PayloadBuilder.java (hunk 25 above)
[the remark this answers is not in the document]
Comment (oliver-walker-hms) [reply src/com/hms/flexyosisoftconnector/PayloadBuilder.java:343]: I missed this one, yeah null was just something random. An empty string would be better

**Comment by oliver-walker-hms** [comment]
@TomKimsey I can change the null string to an empty string and re-push. What are you thoughts on leaving MAX_PAYLOAD_NUM_CHARACTERS = 0?

File: src/com/hms/flexyosisoftconnector/OsisoftJsonPayload.java (hunk 26 above)
Comment (TomKimsey) [reply src/com/hms/flexyosisoftconnector/OsisoftJsonPayload.java:91]: I think better error handling is probably a good idea.

Comment (TomKimsey) [reply src/com/hms/flexyosisoftconnector/OsisoftJsonPayload.java:91]: Im never going to say no to better error handling :)
